File: backend/routes/payments.py
"""
Payment Management Routes
Handles payment records, receipts, and debt payment processing
"""

from flask import Blueprint, request, jsonify
from datetime import datetime, timedelta
from psycopg.rows import dict_row
from functools import wraps
import jwt
import os
from backend.database import get_db_connection

payments_bp = Blueprint('payments', __name__)

# JWT configuration
SECRET_KEY = os.getenv('SECRET_KEY', 'dev-secret-key-change-in-production')

def token_required(f):
    """Decorator to require valid JWT token"""
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        if not token:
            return jsonify({'error': 'Token is missing'}), 401
        
        try:
            if token.startswith('Bearer '):
                token = token[7:]
            data = jwt.decode(token, SECRET_KEY, algorithms=['HS256'])
            request.user_id = data['user_id']
            request.role = data.get('role', 'operator')
        except Exception as e:
            return jsonify({'error': 'Invalid token'}), 401
        
        return f(*args, **kwargs)
    return decorated

def admin_required(f):
    """Decorator to require admin role"""
    @wraps(f)
    @token_required
    def decorated(*args, **kwargs):
        if request.role != 'admin':
            return jsonify({'error': 'Admin access required'}), 403
        return f(*args, **kwargs)
    return decorated


@payments_bp.route('/api/payments', methods=['POST'])
@token_required
def create_payment():
    """Create a new payment record"""
    try:
        data = request.get_json()
        
        # Validate required fields
        required_fields = ['debt_id', 'amount', 'payment_method']
        if not all(field in data for field in required_fields):
            return jsonify({'error': 'Missing required fields'}), 400
        
        debt_id = data['debt_id']
        amount = float(data['amount'])
        payment_method = data['payment_method']
        notes = data.get('notes', '')
        
        # Validate payment method
        valid_methods = ['cash', 'card', 'transfer', 'oxxo', 'bank_deposit']
        if payment_method not in valid_methods:
            return jsonify({'error': f'Invalid payment method. Must be one of: {", ".join(valid_methods)}'}), 400
        
        if amount <= 0:
            return jsonify({'error': 'Amount must be greater than 0'}), 400
        
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            # Check if debt exists and get details
            cur.execute("""
                SELECT id, user_id, total_amount, amount_paid, payment_status
                FROM debts WHERE id = %s
            """, (debt_id,))
            debt = cur.fetchone()
            
            if not debt:
                return jsonify({'error': 'Debt not found'}), 404
            
            remaining = debt['total_amount'] - debt['amount_paid']
            if amount > remaining:
                return jsonify({'error': f'Payment amount exceeds remaining balance of {remaining:.2f}'}), 400
            
            # Generate receipt number
            cur.execute("""
                SELECT generate_receipt_number('standard'::receipt_type) as receipt_number
            """)
            receipt_number = cur.fetchone()['receipt_number']
            
            # Create payment record
            payment_status = 'completed' if payment_method == 'cash' else 'pending'
            processed_date = datetime.now() if payment_method == 'cash' else None
            processed_by = request.user_id if payment_method == 'cash' else None
            
            cur.execute("""
                INSERT INTO payment_records 
                (debt_id, user_id, amount, payment_method, payment_status, 
                 receipt_number, notes, processed_by, processed_date)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
                RETURNING id, payment_date, created_at
            """, (debt_id, debt['user_id'], amount, payment_method, payment_status,
                  receipt_number, notes, processed_by, processed_date))
            
            payment = cur.fetchone()
            payment_id = payment['id']
            
            # Create receipt based on payment method
            if payment_method == 'oxxo':
                cur.execute("""
                    SELECT generate_oxxo_barcode() as barcode
                """)
                barcode = cur.fetchone()['barcode']
                
                cur.execute("""
                    SELECT generate_receipt_number('oxxo'::receipt_type) as receipt_number
                """)
                oxxo_receipt_number = cur.fetchone()['receipt_number']
                
                expiration = datetime.now() + timedelta(days=3)
                
                cur.execute("""
                    INSERT INTO payment_receipts 
                    (payment_id, receipt_type, receipt_number, barcode_data, 
                     amount, expiration_date, instructions)
                    VALUES (%s, 'oxxo', %s, %s, %s, %s, %s)
                    RETURNING id
                """, (payment_id, oxxo_receipt_number, barcode, amount, expiration,
                      'Present this barcode at any OXXO store to complete payment'))
                
            elif payment_method in ['transfer', 'bank_deposit']:
                cur.execute("""
                    SELECT generate_clabe() as clabe
                """)
                clabe = cur.fetchone()['clabe']
                
                cur.execute("""
                    SELECT generate_receipt_number('bank_transfer'::receipt_type) as receipt_number
                """)
                bank_receipt_number = cur.fetchone()['receipt_number']
                
                # Sample bank details
                bank_name = "Banco Ejemplo"
                account_number = "1234567890"
                reference = f"DEBT{debt_id:06d}"
                
                cur.execute("""
                    INSERT INTO payment_receipts 
                    (payment_id, receipt_type, receipt_number, reference_number,
                     bank_name, account_number, clabe, amount, instructions)
                    VALUES (%s, 'bank_transfer', %s, %s, %s, %s, %s, %s, %s)
                    RETURNING id
                """, (payment_id, bank_receipt_number, reference, bank_name, 
                      account_number, clabe, amount,
                      f'Transfer to {bank_name} using reference: {reference}'))
            
            conn.commit()
            
            # Get updated debt info
            cur.execute("""
                SELECT * FROM payment_summary WHERE debt_id = %s
            """, (debt_id,))
            debt_summary = cur.fetchone()
            
            return jsonify({
                'success': True,
                'payment': {
                    'id': payment_id,
                    'receipt_number': receipt_number,
                    'amount': amount,
                    'payment_method': payment_method,
                    'payment_status': payment_status,
                    'payment_date': payment['payment_date'].isoformat() if payment['payment_date'] else None
                },
                'debt_summary': dict(debt_summary) if debt_summary else None
            }), 201
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/payments/<int:debt_id>', methods=['GET'])
@token_required
def get_debt_payments(debt_id):
    """Get all payments for a specific debt"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute("""
                SELECT 
                    pr.*,
                    u.username as processed_by_name
                FROM payment_records pr
                LEFT JOIN users u ON pr.processed_by = u.id
                WHERE pr.debt_id = %s
                ORDER BY pr.payment_date DESC
            """, (debt_id,))
            
            payments = cur.fetchall()
            
            # Get payment summary
            cur.execute("""
                SELECT * FROM payment_summary WHERE debt_id = %s
            """, (debt_id,))
            summary = cur.fetchone()
            
            return jsonify({
                'payments': [dict(p) for p in payments],
                'summary': dict(summary) if summary else None
            }), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/payments/record/<int:payment_id>', methods=['GET'])
@token_required
def get_payment(payment_id):
    """Get a specific payment record"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute("""
                SELECT 
                    pr.*,
                    u.username,
                    d.total_amount as debt_total,
                    p.name as product_name
                FROM payment_records pr
                JOIN users u ON pr.user_id = u.id
                JOIN debts d ON pr.debt_id = d.id
                LEFT JOIN products p ON d.product_id = p.id
                WHERE pr.id = %s
            """, (payment_id,))
            
            payment = cur.fetchone()
            if not payment:
                return jsonify({'error': 'Payment not found'}), 404
            
            return jsonify(dict(payment)), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/payments/record/<int:payment_id>', methods=['PUT'])
@admin_required
def update_payment_status(payment_id):
    """Update payment status (admin only)"""
    try:
        data = request.get_json()
        new_status = data.get('status')
        
        valid_statuses = ['pending', 'processing', 'completed', 'failed', 'cancelled']
        if new_status not in valid_statuses:
            return jsonify({'error': f'Invalid status. Must be one of: {", ".join(valid_statuses)}'}), 400
        
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            # Update payment status
            cur.execute("""
                UPDATE payment_records
                SET payment_status = %s,
                    processed_by = %s,
                    processed_date = CASE WHEN %s = 'completed' THEN CURRENT_TIMESTAMP ELSE processed_date END,
                    updated_at = CURRENT_TIMESTAMP
                WHERE id = %s
                RETURNING *
            """, (new_status, request.user_id, new_status, payment_id))
            
            payment = cur.fetchone()
            if not payment:
                return jsonify({'error': 'Payment not found'}), 404
            
            conn.commit()
            
            return jsonify({
                'success': True,
                'payment': dict(payment)
            }), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/payments/<int:payment_id>/receipt', methods=['GET'])
@token_required
def get_payment_receipt(payment_id):
    """Get receipt information for a payment"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute("""
                SELECT 
                    pr.*,
                    prec.*,
                    p.amount as payment_amount,
                    p.payment_method,
                    p.receipt_number as payment_receipt_number,
                    u.username,
                    u.email,
                    d.total_amount as debt_total
                FROM payment_records p
                LEFT JOIN payment_receipts prec ON p.id = prec.payment_id
                JOIN users u ON p.user_id = u.id
                JOIN debts d ON p.debt_id = d.id
                WHERE p.id = %s
            """, (payment_id,))
            
            receipt = cur.fetchone()
            if not receipt:
                return jsonify({'error': 'Receipt not found'}), 404
            
            return jsonify(dict(receipt)), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/debts/aging', methods=['GET'])
@token_required
def get_aging_report():
    """Get debt aging report"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            # Get aging report
            cur.execute("""
                SELECT * FROM debt_aging_report
                ORDER BY days_overdue DESC, total_amount DESC
            """)
            aging_data = cur.fetchall()
            
            # Get summary by aging category
            cur.execute("""
                SELECT 
                    aging_category,
                    COUNT(*) as debt_count,
                    SUM(balance) as total_balance,
                    AVG(days_overdue) as avg_days_overdue
                FROM debt_aging_report
                GROUP BY aging_category
                ORDER BY 
                    CASE aging_category
                        WHEN 'Current' THEN 1
                        WHEN '1-30 Days' THEN 2
                        WHEN '31-60 Days' THEN 3
                        WHEN '61-90 Days' THEN 4
                        WHEN '90+ Days' THEN 5
                        WHEN 'Paid' THEN 6
                        ELSE 7
                    END
            """)
            summary = cur.fetchall()
            
            # Get total overdue amount
            cur.execute("""
                SELECT 
                    COUNT(*) as total_overdue_count,
                    COALESCE(SUM(balance), 0) as total_overdue_amount
                FROM debt_aging_report
                WHERE aging_category NOT IN ('Current', 'Paid', 'No Due Date')
            """)
            totals = cur.fetchone()
            
            return jsonify({
                'aging_data': [dict(row) for row in aging_data],
                'summary': [dict(row) for row in summary],
                'totals': dict(totals)
            }), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/debts/overdue', methods=['GET'])
@token_required
def get_overdue_debts():
    """Get all overdue debts"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute("""
                SELECT * FROM overdue_debts
            """)
            
            overdue = cur.fetchall()
            
            return jsonify({
                'overdue_debts': [dict(row) for row in overdue],
                'count': len(overdue)
            }), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@payments_bp.route('/api/payments/methods', methods=['GET'])
def get_payment_methods():
    """Get available payment methods"""
    return jsonify({
        'payment_methods': [
            {
                'id': 'cash',
                'name': 'Cash',
                'description': 'Pay in cash at the warehouse',
                'requires_receipt': False,
                'processing_time': 'Immediate'
            },
            {
                'id': 'card',
                'name': 'Credit/Debit Card',
                'description': 'Pay with credit or debit card',
                'requires_receipt': False,
                'processing_time': 'Immediate'
            },
            {
                'id': 'transfer',
                'name': 'Bank Transfer',
                'description': 'Transfer to company bank account',
                'requires_receipt': True,
                'processing_time': '1-2 business days'
            },
            {
                'id': 'oxxo',
                'name': 'OXXO Pay',
                'description': 'Pay at any OXXO store',
                'requires_receipt': True,
                'processing_time': 'Up to 48 hours'
            },
            {
                'id': 'bank_deposit',
                'name': 'Bank Deposit',
                'description': 'Deposit directly to bank account',
                'requires_receipt': True,
                'processing_time': '1-2 business days'
            }
        ]
    }), 200


@payments_bp.route('/api/payments/stats', methods=['GET'])
@admin_required
def get_payment_stats():
    """Get payment statistics (admin only)"""
    try:
        conn = get_db_connection()
        with conn.cursor(row_factory=dict_row) as cur:
            # Total payments by method
            cur.execute("""
                SELECT 
                    payment_method,
                    COUNT(*) as count,
                    SUM(amount) as total_amount,
                    AVG(amount) as avg_amount
                FROM payment_records
                WHERE payment_status = 'completed'
                GROUP BY payment_method
                ORDER BY total_amount DESC
            """)
            by_method = cur.fetchall()
            
            # Payments by status
            cur.execute("""
                SELECT 
                    payment_status,
                    COUNT(*) as count,
                    SUM(amount) as total_amount
                FROM payment_records
                GROUP BY payment_status
            """)
            by_status = cur.fetchall()
            
            # Recent payments
            cur.execute("""
                SELECT 
                    pr.*,
                    u.username,
                    d.total_amount as debt_total
                FROM payment_records pr
                JOIN users u ON pr.user_id = u.id
                JOIN debts d ON pr.debt_id = d.id
                ORDER BY pr.payment_date DESC
                LIMIT 10
            """)
            recent = cur.fetchall()
            
            # Monthly totals
            cur.execute("""
                SELECT 
                    TO_CHAR(payment_date, 'YYYY-MM') as month,
                    COUNT(*) as payment_count,
                    SUM(amount) as total_amount
                FROM payment_records
                WHERE payment_status = 'completed'
                    AND payment_date >= CURRENT_DATE - INTERVAL '12 months'
                GROUP BY TO_CHAR(payment_date, 'YYYY-MM')
                ORDER BY month DESC
            """)
            monthly = cur.fetchall()
            
            return jsonify({
                'by_method': [dict(row) for row in by_method],
                'by_status': [dict(row) for row in by_status],
                'recent_payments': [dict(row) for row in recent],
                'monthly_totals': [dict(row) for row in monthly]
            }), 200
            
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: frontend/assets/js/payments.js
/**
 * Payment Management System
 * Handles payment creation, tracking, receipts, and aging reports
 */

let selectedDebt = null;
let selectedPaymentMethod = null;
let paymentMethods = [];

// Initialize on page load
document.addEventListener('DOMContentLoaded', function() {
    checkAuth();
    loadPaymentMethods();
    loadDebts();
    
    // Set up form submission
    document.getElementById('paymentForm').addEventListener('submit', createPayment);
    
    // Set up debt selection
    document.getElementById('debtSelect').addEventListener('change', handleDebtSelection);
    
    // Load data for other tabs when they're shown
    document.getElementById('payment-history-tab').addEventListener('shown.bs.tab', loadAllPayments);
    document.getElementById('aging-report-tab').addEventListener('shown.bs.tab', loadAgingReport);
    document.getElementById('overdue-tab').addEventListener('shown.bs.tab', loadOverdueDebts);
});

// Load available payment methods
async function loadPaymentMethods() {
    try {
        const response = await fetch('/api/payments/methods');
        const data = await response.json();
        
        paymentMethods = data.payment_methods;
        displayPaymentMethods();
        displayPaymentMethodsInfo();
    } catch (error) {
        console.error('Error loading payment methods:', error);
    }
}

// Display payment method cards
function displayPaymentMethods() {
    const container = document.getElementById('paymentMethods');
    container.innerHTML = '';
    
    paymentMethods.forEach(method => {
        const icon = {
            'cash': 'cash-coin',
            'card': 'credit-card',
            'transfer': 'bank',
            'oxxo': 'shop',
            'bank_deposit': 'building'
        }[method.id] || 'cash';
        
        const card = document.createElement('div');
        card.className = 'col-6';
        card.innerHTML = `
            <div class="payment-method-card card text-center p-3" data-method="${method.id}">
                <i class="bi bi-${icon} fs-1 mb-2"></i>
                <h6>${method.name}</h6>
                <small class="text-muted">${method.processing_time}</small>
            </div>
        `;
        
        card.querySelector('.payment-method-card').addEventListener('click', () => selectPaymentMethod(method.id));
        container.appendChild(card);
    });
}

// Display payment methods info
function displayPaymentMethodsInfo() {
    const container = document.getElementById('paymentMethodsInfo');
    container.innerHTML = paymentMethods.map(method => `
        <div class="mb-3">
            <h6><i class="bi bi-${getMethodIcon(method.id)}"></i> ${method.name}</h6>
            <p class="text-muted small mb-1">${method.description}</p>
            <small><strong>Processing:</strong> ${method.processing_time}</small>
        </div>
        <hr>
    `).join('');
}

// Get icon for payment method
function getMethodIcon(methodId) {
    const icons = {
        'cash': 'cash-coin',
        'card': 'credit-card',
        'transfer': 'bank',
        'oxxo': 'shop',
        'bank_deposit': 'building'
    };
    return icons[methodId] || 'cash';
}

// Select payment method
function selectPaymentMethod(methodId) {
    selectedPaymentMethod = methodId;
    
    // Update UI
    document.querySelectorAll('.payment-method-card').forEach(card => {
        card.classList.remove('selected');
    });
    
    document.querySelector(`[data-method="${methodId}"]`)?.classList.add('selected');
}

// Load pending debts
async function loadDebts() {
    try {
        const response = await apiRequest('/api/debts');
        const debts = response.debts || [];
        
        // Filter for debts that aren't fully paid
        const pendingDebts = debts.filter(debt => 
            debt.payment_status !== 'paid' && 
            (debt.total_amount - (debt.amount_paid || 0)) > 0
        );
        
        const select = document.getElementById('debtSelect');
        select.innerHTML = '<option value="">Select a debt...</option>';
        
        pendingDebts.forEach(debt => {
            const remaining = debt.total_amount - (debt.amount_paid || 0);
            const option = document.createElement('option');
            option.value = debt.id;
            option.textContent = `Debt #${debt.id} - ${debt.username} - $${remaining.toFixed(2)} remaining`;
            option.dataset.debt = JSON.stringify(debt);
            select.appendChild(option);
        });
    } catch (error) {
        console.error('Error loading debts:', error);
        showAlert('Error loading debts', 'danger');
    }
}

// Handle debt selection
function handleDebtSelection(e) {
    const select = e.target;
    const option = select.options[select.selectedIndex];
    
    if (!option.dataset.debt) {
        selectedDebt = null;
        document.getElementById('debtDetails').classList.add('d-none');
        document.getElementById('remainingBalance').textContent = '';
        return;
    }
    
    selectedDebt = JSON.parse(option.dataset.debt);
    const remaining = selectedDebt.total_amount - (selectedDebt.amount_paid || 0);
    
    // Show debt details
    const detailsDiv = document.getElementById('debtDetails');
    detailsDiv.innerHTML = `
        <strong>Debt Details:</strong><br>
        User: ${selectedDebt.username}<br>
        Product: ${selectedDebt.product_name || 'N/A'}<br>
        Total Amount: $${selectedDebt.total_amount.toFixed(2)}<br>
        Paid: $${(selectedDebt.amount_paid || 0).toFixed(2)}<br>
        <strong>Remaining: $${remaining.toFixed(2)}</strong>
    `;
    detailsDiv.classList.remove('d-none');
    
    // Set max payment amount
    document.getElementById('paymentAmount').max = remaining;
    document.getElementById('paymentAmount').value = remaining.toFixed(2);
    document.getElementById('remainingBalance').textContent = `Remaining balance: $${remaining.toFixed(2)}`;
}

// Create payment
async function createPayment(e) {
    e.preventDefault();
    
    if (!selectedDebt) {
        showAlert('Please select a debt', 'warning');
        return;
    }
    
    if (!selectedPaymentMethod) {
        showAlert('Please select a payment method', 'warning');
        return;
    }
    
    const amount = parseFloat(document.getElementById('paymentAmount').value);
    const notes = document.getElementById('paymentNotes').value;
    
    try {
        const response = await apiRequest('/api/payments', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                debt_id: selectedDebt.id,
                amount: amount,
                payment_method: selectedPaymentMethod,
                notes: notes
            })
        });
        
        showAlert('Payment created successfully!', 'success');
        
        // Show receipt if applicable
        if (['oxxo', 'transfer', 'bank_deposit'].includes(selectedPaymentMethod)) {
            showReceipt(response.payment.id);
        }
        
        // Reset form
        document.getElementById('paymentForm').reset();
        selectedDebt = null;
        selectedPaymentMethod = null;
        document.querySelectorAll('.payment-method-card').forEach(card => {
            card.classList.remove('selected');
        });
        document.getElementById('debtDetails').classList.add('d-none');
        
        // Reload debts
        loadDebts();
        
    } catch (error) {
        console.error('Error creating payment:', error);
        showAlert(error.message || 'Error creating payment', 'danger');
    }
}

// Show receipt
async function showReceipt(paymentId) {
    try {
        const response = await apiRequest(`/api/payments/${paymentId}/receipt`);
        
        const modal = new bootstrap.Modal(document.getElementById('receiptModal'));
        const content = document.getElementById('receiptContent');
        
        if (response.receipt_type === 'oxxo') {
            content.innerHTML = generateOxxoReceipt(response);
        } else if (response.receipt_type === 'bank_transfer') {
            content.innerHTML = generateBankReceipt(response);
        } else {
            content.innerHTML = generateStandardReceipt(response);
        }
        
        modal.show();
    } catch (error) {
        console.error('Error loading receipt:', error);
    }
}

// Generate OXXO receipt
function generateOxxoReceipt(data) {
    return `
        <div class="receipt-container text-center">
            <h3><i class="bi bi-shop"></i> OXXO Pay</h3>
            <hr>
            <p><strong>Receipt Number:</strong> ${data.receipt_number}</p>
            <p><strong>Amount to Pay:</strong> <span class="fs-4 text-primary">$${data.amount.toFixed(2)}</span></p>
            <p><strong>Reference:</strong> ${data.reference_number}</p>
            
            <div class="my-4">
                <p class="text-muted">Present this barcode at any OXXO store</p>
                <div class="barcode mx-auto" style="max-width: 400px;">
                    ${data.barcode_data}
                </div>
            </div>
            
            <div class="alert alert-warning">
                <i class="bi bi-clock"></i> <strong>Valid until:</strong> ${formatDate(data.expiration_date)}
            </div>
            
            <p class="text-muted small">${data.instructions}</p>
        </div>
    `;
}

// Generate bank transfer receipt
function generateBankReceipt(data) {
    return `
        <div class="receipt-container">
            <h3><i class="bi bi-bank"></i> Bank Transfer Details</h3>
            <hr>
            <div class="row">
                <div class="col-md-6">
                    <p><strong>Receipt Number:</strong><br>${data.receipt_number}</p>
                    <p><strong>Amount to Transfer:</strong><br><span class="fs-4 text-primary">$${data.amount.toFixed(2)}</span></p>
                </div>
                <div class="col-md-6">
                    <p><strong>Bank:</strong><br>${data.bank_name}</p>
                    <p><strong>Account Number:</strong><br>${data.account_number}</p>
                </div>
            </div>
            
            <div class="alert alert-info">
                <p class="mb-2"><strong>CLABE (Interbank Key):</strong></p>
                <p class="clabe-input fs-5 text-center mb-0">${data.clabe}</p>
            </div>
            
            <div class="alert alert-warning">
                <strong><i class="bi bi-info-circle"></i> Reference Number:</strong> ${data.reference_number}
                <p class="mb-0 mt-2 small">Include this reference in your transfer to ensure proper credit</p>
            </div>
            
            <p class="text-muted small">${data.instructions}</p>
        </div>
    `;
}

// Generate standard receipt
function generateStandardReceipt(data) {
    return `
        <div class="receipt-container">
            <h3><i class="bi bi-receipt"></i> Payment Receipt</h3>
            <hr>
            <p><strong>Receipt Number:</strong> ${data.payment_receipt_number}</p>
            <p><strong>Payment Method:</strong> ${data.payment_method.toUpperCase()}</p>
            <p><strong>Amount:</strong> $${data.payment_amount.toFixed(2)}</p>
            <p><strong>User:</strong> ${data.username}</p>
            <p><strong>Email:</strong> ${data.email}</p>
            <hr>
            <p class="text-center text-success"><i class="bi bi-check-circle"></i> Payment Completed</p>
        </div>
    `;
}

// Load all payments
async function loadAllPayments() {
    try {
        // This would need a backend endpoint to get all payments
        // For now, show a placeholder
        const tbody = document.querySelector('#paymentsTable tbody');
        tbody.innerHTML = '<tr><td colspan="8" class="text-center">Payment history coming soon...</td></tr>';
    } catch (error) {
        console.error('Error loading payments:', error);
    }
}

// Load aging report
async function loadAgingReport() {
    try {
        const response = await apiRequest('/api/debts/aging');
        
        displayAgingSummary(response.summary, response.totals);
        displayAgingTable(response.aging_data);
    } catch (error) {
        console.error('Error loading aging report:', error);
        showAlert('Error loading aging report', 'danger');
    }
}

// Display aging summary cards
function displayAgingSummary(summary, totals) {
    const container = document.getElementById('agingSummaryCards');
    container.innerHTML = '';
    
    summary.forEach(category => {
        const card = document.createElement('div');
        card.className = 'col-12 mb-3';
        card.innerHTML = `
            <div class="card">
                <div class="card-body">
                    <h6 class="card-title">${category.aging_category}</h6>
                    <p class="mb-1"><strong>${category.debt_count}</strong> debts</p>
                    <p class="mb-0 text-success"><strong>$${parseFloat(category.total_balance || 0).toFixed(2)}</strong></p>
                </div>
            </div>
        `;
        container.appendChild(card);
    });
    
    // Add totals card
    const totalsCard = document.createElement('div');
    totalsCard.className = 'col-12 mb-3';
    totalsCard.innerHTML = `
        <div class="card bg-danger text-white">
            <div class="card-body">
                <h6 class="card-title">Total Overdue</h6>
                <p class="mb-1"><strong>${totals.total_overdue_count}</strong> debts</p>
                <p class="mb-0"><strong>$${parseFloat(totals.total_overdue_amount || 0).toFixed(2)}</strong></p>
            </div>
        </div>
    `;
    container.appendChild(totalsCard);
}

// Display aging table
function displayAgingTable(data) {
    const tbody = document.querySelector('#agingTable tbody');
    tbody.innerHTML = '';
    
    data.forEach(row => {
        const tr = document.createElement('tr');
        tr.innerHTML = `
            <td>${row.id}</td>
            <td>${row.username}</td>
            <td>${row.product_name || 'N/A'}</td>
            <td>$${parseFloat(row.total_amount).toFixed(2)}</td>
            <td>$${parseFloat(row.balance).toFixed(2)}</td>
            <td>${row.payment_due_date ? formatDate(row.payment_due_date) : 'N/A'}</td>
            <td>${row.days_overdue}</td>
            <td><span class="badge aging-badge ${getAgingClass(row.aging_category)}">${row.aging_category}</span></td>
        `;
        tbody.appendChild(tr);
    });
}

// Get aging class for styling
function getAgingClass(category) {
    const classes = {
        'Current': 'aging-current',
        '1-30 Days': 'aging-1-30',
        '31-60 Days': 'aging-31-60',
        '61-90 Days': 'aging-61-90',
        '90+ Days': 'aging-90-plus'
    };
    return classes[category] || '';
}

// Load overdue debts
async function loadOverdueDebts() {
    try {
        const response = await apiRequest('/api/debts/overdue');
        
        const tbody = document.querySelector('#overdueTable tbody');
        tbody.innerHTML = '';
        
        response.overdue_debts.forEach(debt => {
            const tr = document.createElement('tr');
            tr.innerHTML = `
                <td>${debt.id}</td>
                <td>${debt.username}</td>
                <td>${debt.email}</td>
                <td>${debt.product_name || 'N/A'}</td>
                <td class="text-danger">$${parseFloat(debt.balance).toFixed(2)}</td>
                <td>${formatDate(debt.payment_due_date)}</td>
                <td><span class="badge bg-danger">${debt.days_overdue} days</span></td>
                <td>
                    <button class="btn btn-sm btn-primary" onclick="createPaymentForDebt(${debt.id})">
                        <i class="bi bi-cash"></i> Pay
                    </button>
                </td>
            `;
            tbody.appendChild(tr);
        });
        
        if (response.overdue_debts.length === 0) {
            tbody.innerHTML = '<tr><td colspan="8" class="text-center text-success">No overdue debts! 🎉</td></tr>';
        }
    } catch (error) {
        console.error('Error loading overdue debts:', error);
        showAlert('Error loading overdue debts', 'danger');
    }
}

// Create payment for specific debt
function createPaymentForDebt(debtId) {
    // Switch to new payment tab
    const tab = new bootstrap.Tab(document.getElementById('new-payment-tab'));
    tab.show();
    
    // Select the debt
    document.getElementById('debtSelect').value = debtId;
    document.getElementById('debtSelect').dispatchEvent(new Event('change'));
}

// Print receipt
function printReceipt() {
    const content = document.getElementById('receiptContent').innerHTML;
    const printWindow = window.open('', '', 'height=600,width=800');
    printWindow.document.write('<html><head><title>Receipt</title>');
    printWindow.document.write('<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet">');
    printWindow.document.write('<style>.receipt-container { margin: 20px; } .barcode { font-family: monospace; font-size: 24px; }</style>');
    printWindow.document.write('</head><body>');
    printWindow.document.write(content);
    printWindow.document.write('</body></html>');
    printWindow.document.close();
    printWindow.print();
}

// Format date
function formatDate(dateString) {
    if (!dateString) return 'N/A';
    const date = new Date(dateString);
    return date.toLocaleDateString();
}

// Show alert
function showAlert(message, type = 'info') {
    const alertDiv = document.createElement('div');
    alertDiv.className = `alert alert-${type} alert-dismissible fade show position-fixed top-0 start-50 translate-middle-x mt-3`;
    alertDiv.style.zIndex = '9999';
    alertDiv.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
    document.body.appendChild(alertDiv);
    
    setTimeout(() => alertDiv.remove(), 5000);
}

// API request helper (should match your existing auth.js)
async function apiRequest(url, options = {}) {
    const token = localStorage.getItem('flowrack_token');
    if (!token) {
        window.location.href = 'index.html';
        throw new Error('Not authenticated');
    }
    
    const defaultOptions = {
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json'
        }
    };
    
    const response = await fetch(url, { ...defaultOptions, ...options });
    
    if (response.status === 401) {
        localStorage.removeItem('flowrack_token');
        window.location.href = 'index.html';
        throw new Error('Session expired');
    }
    
    if (!response.ok) {
        const error = await response.json();
        throw new Error(error.error || 'Request failed');
    }
    
    return response.json();
}

// Check authentication
function checkAuth() {
    const token = localStorage.getItem('flowrack_token');
    if (!token) {
        window.location.href = 'index.html';
        return;
    }
    
    // Display user info
    const user = JSON.parse(localStorage.getItem('flowrack_user') || '{}');
    document.getElementById('userDisplay').textContent = user.username || 'User';
}
